os/arch/arm/src/imxrt/imxrt_i2s.c

```diff
@@ -0,0 +1,2707 @@
+		DEBUGASSERT(ret == 0 || errno == EINTR);
+	} while (ret < 0);
+}
+
+/****************************************************************************
+ * Name: i2s_rxdatawidth
+ *
+ * Description:
+ *   Set the I2S RX data width.  The RX bitrate is determined by
+ *   sample_rate * data_width.
+ *
+ * Input Parameters:
+ *   dev   - Device-specific state data
+ *   width - The I2S data with in bits.
+ *
+ * Returned Value:
+ *   Returns the resulting bitrate
+ *
+ ****************************************************************************/
+
+static uint32_t i2s_rxdatawidth(struct i2s_dev_s *dev, int bits)
+{
+#if defined(I2S_HAVE_RX) && (0 < I2S_HAVE_RX)
+	struct imxrt_i2s_s *priv = (struct imxrt_i2s_s *)dev;
+	int ret;
+
+	/* Support 16/32bits only */
+	DEBUGASSERT(priv && (bits == I2S_BITS_PER_SAMPLE_16BIT || bits == I2S_BITS_PER_SAMPLE_32BIT));
+
+	if (bits == I2S_BITS_PER_SAMPLE_16BIT || bits == I2S_BITS_PER_SAMPLE_32BIT) {
+		priv->bits_per_sample = bits;
+		ret = i2s_set_clk(priv, priv->sample_rate, priv->bits_per_sample, priv->channel_num);
+		if (ret == OK) {
+			return priv->bits_per_sample * priv->sample_rate;
+		}
+	}
+#endif
+
+	return 0;
+}
+
+/****************************************************************************
+ * Name: i2s_receive
+ *
+ * Description:
+ *   Receive a block of data from I2S.
+ *
+ * Input Parameters:
+ *   dev      - Device-specific state data
+ *   apb      - A pointer to the audio buffer in which to recieve data
+ *   callback - A user provided callback function that will be called at
+ *              the completion of the transfer.  The callback will be
+ *              performed in the context of the worker thread.
+ *   arg      - An opaque argument that will be provided to the callback
+ *              when the transfer complete
+ *   timeout  - The timeout value to use.  The transfer will be canceled
+ *              and an ETIMEDOUT error will be reported if this timeout
+ *              elapsed without completion of the DMA transfer.  Units
+ *              are system clock ticks.  Zero means no timeout.
+ *
+ * Returned Value:
+ *   OK on success; a negated errno value on failure.  NOTE:  This function
+ *   only enqueues the transfer and returns immediately.  Success here only
+ *   means that the transfer was enqueued correctly.
+ *
+ *   When the transfer is complete, a 'result' value will be provided as
+ *   an argument to the callback function that will indicate if the transfer
+ *   failed.
+ *
+ ****************************************************************************/
+
+static int i2s_receive(struct i2s_dev_s *dev, struct ap_buffer_s *apb, i2s_callback_t callback, void *arg, uint32_t timeout)
+{
+	struct imxrt_i2s_s *priv = (struct imxrt_i2s_s *)dev;
+
+	i2sinfo("[I2S RX] apb=%p nmaxbytes=%d samp=%p arg=%p timeout=%d\n", apb, apb->nmaxbytes, apb->samp, arg, timeout);
+
+#if defined(I2S_HAVE_RX) && (0 < I2S_HAVE_RX)
+	/* Has the RX channel been enabled? */

```

Comment:
Internally, does i2s maintain a single linear buffer? or is it a pipelined buffer?

Reply:
In case of imxrt_i2s.c, we are using 30 linear buffers.
But, as I said above, we plan to use imxrt_i2s_edma.c file including eDMA function